### DISCUSSION TITLE
Potentially returning too many overloads?

### DISCUSSION THREAD

**Discussion by rchiodo** [microsoft/pyright#6964] [Q&A] [answered]
@erictraut, We (pylance) have a test for (fourslash) code like this:

```
// @filename: test.py
//// from decl import O
////
//// O().[|/*marker*/foo|]

// @filename: decl.py
//// from typing import overload
////
//// class [|/*overload*/O|]:
////    @overload
////    def foo(self): pass
////    @overload
////    def foo(self, a: int): pass
////    def foo(self, a: int | None): pass
```

The test is supposed to verify that we return the correct number of declarations in an overload situation. The test is passing currently. 

After pulling the latest Pyright, the test fails because the number of overloads has changed from 2 to 3.

This change here makes it return 3 instead:
https://github.com/microsoft/pyright/commit/0674b096fc21fad82765dd7084aa6dee0eea49d6

I'm not sure if this is incorrect or not. It seems like there's 3 overloads here, so it seems like our test is wrong, but wanted to double check.

**Comment by rchiodo** [comment]
It's this specific change here:

```
// Add the unmodified implementation if it's present.
    const implementation = OverloadedFunctionType.getImplementation(type);
    if (implementation) {
        newSignatures.push(implementation);
    }
``` 

The old code would skip the implementation in the new signatures list.

**Reply by erictraut** [reply, marked as answer]
The `OverloadedFunctionType` retains a list of overloads plus the implementation. Previously, I ditched the implementation when binding a method, but I recently needed to change that. So now the implementation is retained. The signature help provider should use only the overloads and ignore the implementation. You can use the `OverloadedFunctionType.getOverloads()` method to retrieve them.

Note that my recent change affected only methods, not regular functions. So if the signature help provider wasn't filtering for overloads, then this was a long-standing bug. You might want to update the test to cover that case as well.
